pkg/command/plugin.go

```diff
@@ -441,187 +440,131 @@ func syncPlugins(cmd *cobra.Command) error {
+		}
+		if p.recommended != "" && p.installed != p.recommended {
+			p.status = pluginStatusRecommendUpdate
+			pluginSyncRequired = true
+		}
+		plugins = append(plugins, p)
+	}
+	for index := range recommendedContextPlugins {
+		if recommendedContextPlugins[index].Status != common.PluginStatusInstalled {
+			p := pluginListInfo{
+				name:        recommendedContextPlugins[index].Name,
+				description: recommendedContextPlugins[index].Description,
+				target:      string(recommendedContextPlugins[index].Target),
+				installed:   "",
+				recommended: recommendedContextPlugins[index].RecommendedVersion,
+				status:      pluginStatusRecommendInstall,
+				contextName: recommendedContextPlugins[index].ContextName,
+				active:      false,
 			}
-			outputWriter.AddRow(
-				ctxSpecificPlugins[i].Name,
-				ctxSpecificPlugins[i].Description,
-				string(ctxSpecificPlugins[i].Target),
-				v,
-				ctxSpecificPlugins[i].Status,
-			)
+			plugins = append(plugins, p)
+			pluginSyncRequired = true
 		}
-		outputWriter.Render()
 	}
 
-	if pluginSyncRequired {
-		// Print a warning to the user that some context plugins are not installed or outdated and plugin sync is required to install them
-		fmt.Println("")
-		fmt.Println("Note: As shown above, some recommended plugins have not been installed or are outdated. To install them please run 'tanzu plugin sync'.")
-	}
-}
+	sort.Sort(pluginListInfoSorter(plugins))
 
-func displayInstalledAndMissingListView(installedStandalonePlugins []cli.PluginInfo, installedContextPlugins, missingContextPlugins []discovery.Discovered, writer io.Writer) {
-	// List installed standalone plugins
-	outputWriter := component.NewOutputWriterWithOptions(writer, outputFormat, []component.OutputWriterOption{}, "Name", "Description", "Target", "Version", "Status", "Context")
-	sort.Sort(cli.PluginInfoSorter(installedStandalonePlugins))
-	for index := range installedStandalonePlugins {
-		outputWriter.AddRow(
-			installedStandalonePlugins[index].Name,
-			installedStandalonePlugins[index].Description,
-			string(installedStandalonePlugins[index].Target),
-			installedStandalonePlugins[index].Version,
-			installedStandalonePlugins[index].Status,
-			"", // No context
-		)
+	outputPluginWriter := component.NewOutputWriterWithOptions(writer, outputFormat, []component.OutputWriterOption{})
+	if isTableOutputFormat() {
+		columnsNames := []string{"Name", "Description", "Target", "Installed", "Recommended", "Status"}
+		if showAllColumns {
+			columnsNames = append(columnsNames, "Active")
+		}
+		outputPluginWriter.SetKeys(columnsNames...)
+		outputPluginWriter.MarkDynamicKeys("Recommended") // Marking this column as dynamic so that it will only be shown if at least one row is non-empty
+		for index := range plugins {
+			outputPluginWriter.AddRow(plugins[index].name, plugins[index].description, plugins[index].target, plugins[index].installed, plugins[index].recommended, plugins[index].status, plugins[index].active)
```

Comment:
Are we relying on the tail end of the row values being ignored when showAllColumns is not set?
Can't recall if there were precendence of this; if not; it might warrant a small comment.

Reply:
Added comment.